File: src/day14.py
import math
import re
from collections import Counter
from dataclasses import dataclass


@dataclass
class Robot:
    px: int
    py: int
    vx: int
    vy: int

    def tick(self):
        self.px += self.vx
        self.py += self.vy
        self.px %= WIDTH
        self.py %= HEIGHT


WIDTH = 101
HEIGHT = 103
MID_WIDTH = WIDTH // 2
MID_HEIGHT = HEIGHT // 2


def count_quadrants(counts):
    quads = [0, 0, 0, 0]
    for pt, num in counts.items():
        x, y = pt
        quad = 2 * (y > MID_HEIGHT) + int(x > MID_WIDTH)
        quads[quad] += num
    return quads


def display(counts, f=None):
    for r in range(HEIGHT):
        for c in range(WIDTH):
            if c == MID_WIDTH or r == MID_HEIGHT:
                print(" ", end="", file=f)
                continue
            c = counts[(c, r)]
            print(c if c > 0 else ".", end="", file=f)
        print(file=f)
    print(file=f)


def solve(lines):
    final_positions = []
    for robot in parse_robots(lines):
        fx = (robot.px + 100 * robot.vx) % WIDTH
        fy = (robot.py + 100 * robot.vy) % HEIGHT
        if fx != MID_WIDTH and fy != MID_HEIGHT:
            final_positions.append((fx, fy))

    counts = Counter(final_positions)
    return math.prod(count_quadrants(counts))


def simulate(robots: list[Robot], write_to_file=False):
    easter_egg_frame = 0
    min_safety_factor = 1000000000
    max_frames = 10_000
    max_counts = None
    frame = 1

    while frame < max_frames:
        for robot in robots:
            robot.px += robot.vx
            robot.px %= WIDTH
            robot.py += robot.vy
            robot.py %= HEIGHT
        counts = Counter([(r.px, r.py) for r in robots])

        safety_factor = math.prod(count_quadrants(counts))
        if safety_factor < min_safety_factor:
            easter_egg_frame = frame
            min_safety_factor = safety_factor
            max_counts = counts

        frame += 1

    if write_to_file:
        with open("tree_frame.txt", "w") as f:
            display(max_counts, f)

    return easter_egg_frame


def parse_robots(lines):
    robots = []
    for line in lines:
        nums = [int(n) for n in re.findall(r"(-?\d+)", line)]
        robots.append(Robot(*nums))
    return robots


def main():
    with open("data/input14.txt") as f:
        s = f.read().split("\n")

    print(f"part 1: {solve(s)}")
    print(f"part 2: {simulate(parse_robots(s))}")


if __name__ == "__main__":
    main()

File: src/day14.zig
const std = @import("std");
const lib = @import("lib.zig");
pub const input = "data/input14.txt";

const WIDTH: i64 = 101;
const HEIGHT: i64 = 103;
const NUM_ROBOTS: usize = 500;

const Robot = struct {
    px: i64 = 0,
    py: i64 = 0,
    vx: i64 = 0,
    vy: i64 = 0,

    const Self = @This();
    fn tick(self: *Self) void {
        self.px = @mod(self.px + self.vx, WIDTH);
        self.py = @mod(self.py + self.vy, HEIGHT);
    }

    fn tickn(self: *Self, n: i64) void {
        self.px = @mod(self.px + self.vx * n, WIDTH);
        self.py = @mod(self.py + self.vy * n, HEIGHT);
    }
};

fn count_quads(robots: []Robot) usize {
    var quads: [4]usize = .{ 0, 0, 0, 0 };
    for (robots) |r| {
        var which: usize = 0;
        if (r.px == WIDTH / 2 or r.py == HEIGHT / 2) continue;
        if (r.px > WIDTH / 2) which += 1;
        if (r.py > HEIGHT / 2) which += 2;
        quads[which] += 1;
    }

    return quads[0] * quads[1] * quads[2] * quads[3];
}

fn find_tree(robots: []Robot, n: usize) usize {
    var frame_num: usize = 1;
    var min_safety_factor: usize = 1_000_000_000_000_000;
    var min_frame: usize = 0;
    while (frame_num < n) : (frame_num += 1) {
        for (0..NUM_ROBOTS) |rn| {
            _ = &robots[rn].tick();
        }
        const safety_factor = count_quads(robots);
        if (safety_factor < min_safety_factor) {
            min_frame = frame_num;
            min_safety_factor = safety_factor;
        }
    }
    return min_frame;
}

fn parse_robots(data: []const u8, alloc: std.mem.Allocator) ![]Robot {
    var start: usize = 0;
    var end: usize = 0;
    var robots = std.ArrayList(Robot).init(alloc);
    for (0..NUM_ROBOTS) |_| {
        var numbers: [4]i64 = undefined;

        for (0..4) |i| {
            start = if (end > 0) end + 1 else 0;
            while (!std.ascii.isDigit(data[start]) and data[start] != '-') {
                start += 1;
            }
            end = start + 1;
            while (end < data.len and std.ascii.isDigit(data[end])) {
                end += 1;
            }

            numbers[i] = try std.fmt.parseInt(i64, data[start..end], 10);
        }

        try robots.append(Robot{
            .px = numbers[0],
            .py = numbers[1],
            .vx = numbers[2],
            .vy = numbers[3],
        });
    }

    return try robots.toOwnedSlice();
}

pub fn solve(rdr: anytype, part: lib.Part) !usize {
    const alloc = std.heap.page_allocator;

    const data = try rdr.readAllAlloc(alloc, 1000000000);
    defer alloc.free(data);

    var robots = try parse_robots(data, alloc);
    defer alloc.free(robots);

    if (part == .one) {
        for (0..NUM_ROBOTS) |n| {
            robots[n].tickn(100);
        }
        return count_quads(robots);
    } else {
        const frame = find_tree(robots, 10_000);
        return frame;
    }

    return 0;
}

pub fn main() !void {
    try lib.run(usize, solve, "Day 14", input);
}
